torch/testing/_internal/compilable_quantized_tensor.py

```diff
@@ -0,0 +1,401 @@
+        kwargs = {} if kwargs is None else kwargs
+
+        if func is aten.detach.default:
+            return return_and_correct_aliasing(
+                func, args, kwargs, args[0]._apply_fn_to_data(torch.detach)
+            )
+
+        raise NotImplementedError(
+            f"MyDTypeLayout dispatch: attempting to run {func}, this is not supported"
+        )
+
+
+#####################################################
+# torch functional and aten operator implementation #
+#####################################################
+
+implements = MyDTypeTensor.implements
+
+
+def _quantized_linear_op(
+    input_tensor: torch.Tensor, weight_tensor: torch.Tensor, bias: torch.Tensor
+) -> torch.Tensor:
+    if isinstance(input_tensor, MyDTypeTensor):
+        input_tensor = input_tensor.dequantize()
+    if isinstance(weight_tensor, MyDTypeTensor):
+        weight_tensor = weight_tensor.dequantize()
+    return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+
+@implements(torch.nn.functional.linear)
+def _(func, types, args, kwargs):  # type: ignore[no-untyped-def]
+    input_tensor, weight_tensor, bias = (
+        args[0],
+        args[1],
+        args[2] if len(args) > 2 else None,
+    )
+    # using try/except here so that we can have a general fallback when input_tensor/weight_tensor
+    # is not picked up by any of the dispatch paths in `_quantized_linear_op`, this allows us to
+    # make the branches easier to understand in `_quantized_linear_op`
+    try:
+        return _quantized_linear_op(input_tensor, weight_tensor, bias)
+    except NotImplementedError:
+        if isinstance(input_tensor, MyDTypeTensor):
+            input_tensor = input_tensor.dequantize()
+        if isinstance(weight_tensor, MyDTypeTensor):
+            weight_tensor = weight_tensor.dequantize()
+        return torch.nn.functional.linear(input_tensor, weight_tensor, bias)
+
+
+@implements(aten.detach.default)
+def _(func, types, args, kwargs):  # type: ignore[no-untyped-def]
+    # `return_and_correct_aliasing` should be used by wrapper tensor ``__torch_dispatch__`` subclasses that would like to
+    # work with torch.compile. It ensures that the subclass properly implements the aliasing behavior of every op,
+    # which is needed for correctness in AOTAutograd.
+
+    # `_apply_fn_to_data` just applies the function to the tensor data in `args[0]`, `args[0]` is a tensor subclass
+    # of `my_dtype`
+    return return_and_correct_aliasing(
+        func, args, kwargs, args[0]._apply_fn_to_data(torch.detach)
+    )
+
+
+class M(torch.nn.Module):
+    def __init__(self) -> None:
+        super().__init__()
+        self.linear = torch.nn.Linear(1024, 1024)
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        return self.linear(x)
+
+
+#####################
+# Factory functions #
+#####################
+to_my_dtype = MyDTypeTensor.from_float
+
+
+########
+# Test #
```

Comment:
Note that test should be in a proper test file, not inline here.
But will let @bdhirsh comment here on how much he wants to see on this PR vs follow ups.

Reply:
sure, please let me know where should I put it